packages/redux-routine/src/index.js

```diff
@@ -0,0 +1,51 @@
+/**
+ * Internal dependencies
+ */
+import isGenerator from './is-generator';
+
+/**
+ * Creates a Redux middleware, given an object of controls where each key is an
+ * action type for which to act upon, the value a function which returns either
+ * a promise which is to resolve when evaluation of the action should continue,
+ * or a value. The value or resolved promise value is assigned on the return
+ * value of the yield assignment. If the control handler returns undefined, the
+ * execution is not continued.
+ *
+ * @param {Object} controls Object of control handlers.
+ *
+ * @return {Function} Redux middleware function.
+ */
+export default function createMiddleware( controls = {} ) {
+	return ( store ) => ( next ) => ( action ) => {
+		if ( ! isGenerator( action ) ) {
+			return next( action );
+		}
+
+		function step( nextAction ) {
+			if ( ! nextAction ) {
+				return;
+			}
+
+			const control = controls[ nextAction.type ];
+			if ( typeof control === 'function' ) {
+				const routine = control( nextAction );
+
+				if ( routine instanceof Promise ) {
+					// Async control routine awaits resolution.
+					routine.then( ( result ) => {
```

Comment:
Why the handling of Promises is built-in into the runtime, it should be just another control.

Reply:
>Why the handling of Promises is built-in into the runtime, it should be just another control.

Promises are the internal mechanism for the delay, the equivalent of `rungen`'s `next` function. I don't see how it's possible to not have this exist the way it does.

To be clear, this doesn't enable a developer to yield promises from their action creator. This only enables a developer to return a promise from the _implementation of their control_.